File: sast-python3/case/completeness/single_app_tracing/expression/special_expression/enum_001_T.py
# -*- coding: utf-8 -*-
# evaluation information start
# real case = true
# evaluation item = 完整度->单应用跟踪完整度->表达式->语言特有的表达式
# scene introduction = 枚举
# level = 2
# bind_url = completeness/single_app_tracing/expression/special_expression/enum_001_T
# evaluation information end

from enum import Enum
import os


def enum_001_T(taint_src):

    #把taint_src的值当成枚举
    class Color(Enum):
      RED = 1
      GREEN = 2
      BLUE = taint_src

    color_map = {"red": Color.RED, "green": Color.GREEN, "blue": Color.BLUE}
    taint_sink(color_map["blue"].value)


def taint_sink(o):
    os.system(str(o))


if __name__ == '__main__':
    taint_src = "taint_src_value"  
    enum_001_T(taint_src)

File: sast-php/case/completeness/single_app_tracing/class/enum_001_T.php
<?php
// evaluation information start
// real case = true
// evaluation item = 完整度->单应用追踪->类与对象
// scene introduction = 枚举方法传递污点
// level = 2
// bind_url = completeness/single_app_tracing/class/enum_001_T
// evaluation information end

enum Status: string {
    case Active = 'active';
    case Inactive = 'inactive';

    public function format($input): string {
        return $input;
    }
}

function enum_001_T($__taint_src) {
    $status = Status::Active;
    $result = $status->format($__taint_src);
    __taint_sink($result);
}

function __taint_sink($o) {
    shell_exec($o);
}

$taint_src = "taint_src_value";
enum_001_T($taint_src);
